src/assign/propagate.rs

```diff
@@ -760,6 +769,299 @@ impl PropagateIF for AssignStack {
+                            self.level[propagating.vi()],
+                        );
+                    }
+                }
+            }
+            //
+            //## normal clause loop
+            //
+            let mut source = cdb.watch_cache_iter(propagating);
+            let transformers = source
+                .clone()
+                .map(|index| cdb.fetch_watch_cache_entry(propagating, index))
+                .map(|(cid, cached)| {
+                    self.build_propagatation_context(false_lit, cid, &cdb[cid], cached)
+                })
+                .collect::<Vec<PropagationContext>>();
+            // let mut source = cdb.watch_cache_iter(propagating);
+            // while let Some((cid, cached)) = source
+            //     .current()
+            //     .map(|index| cdb.fetch_watch_cache_entry(propagating, index))
+            // {
+            for context in transformers.iter() {
+                match *context {
+                    // let cls = &cdb[cid];
+                    // match self.build_propagatation_context(false_lit, cid, cls, cached) {
+                    PropagationContext::Conflict(cid, cached, flip_watches, cache) => {
+                        if flip_watches {
+                            cdb.swap_watch(cid);
+                        }
+                        cdb.transform_by_restoring_watch_cache(propagating, &mut source, cache);
+                        check_in!(cid, Propagate::EmitConflict(self.num_conflict + 1, cached));
+                        conflict_path!(cached, AssignReason::Implication(cid));
+                    }
+                    PropagationContext::Satisfied(_cid, cached) => {
+                        // The following doesn't need an ID but the internal pointer in the iterator
+                        cdb.transform_by_restoring_watch_cache(propagating, &mut source, cached);
+                        check_in!(cid, Propagate::CacheSatisfied(self.num_conflict));
+                    }
+                    PropagationContext::UnitPropagation(cid, cached, flip_watches, cache) => {
+                        if flip_watches {
+                            cdb.swap_watch(cid);
+                        }
+                        cdb.transform_by_restoring_watch_cache(propagating, &mut source, cache);
+                        debug_assert_eq!(cdb[cid].lit0(), cached);
+                        debug_assert_eq!(self.assigned(cached), None);
+                        // debug_assert!(other_watch_value.is_none());
+                        self.assign_by_implication(cached, AssignReason::Implication(cid));
+                        check_in!(cid, Propagate::BecameUnit(self.num_conflict, cached));
+                    }
+                    PropagationContext::UpdateWatch(cid, new_watch, k, false_watch_pos) => {
+                        cdb.detach_watch_cache(propagating, &mut source);
+                        cdb.transform_by_updating_watch(cid, false_watch_pos, k, true);
```

Comment:
🐛  This function reorders the items in watchlist by calling `swap_remove`! 🐛

Reply:
So we have to
1. build a new list during the iteration, then swap to the original.
2. execute in-place shrinking, then resize after the iteration
3. use $O(n)$ `remove`